### PR TITLE
Add a reminder when publishing new toots with media that has no image description

### LINKED ISSUE #19974: Optional mandatory image captions
### Pitch

I am doing my best to always add good image captions. I would love to be better at it. With an option to enforce image captions I would never forget.

### Motivation

Since joining the Fediverse I've been hearing more from users on screen readers who really value the higher rates of image caption compared to other social media. I will keep trying to do better at this, but a technical means to enforce it would help.

### PR DESCRIPTION
**Pull request by connorshea**
(Redo of #21654 because I was having problems getting Circle to run)

This resolves #19974 and #7804.

Short demo video via YouTube because the file is too big for GitHub: https://www.youtube.com/watch?v=ZBn6Z7-jej8

Essentially, this does two main things:

- Adds a setting under "Preferences > Other" for enabling this reminder dialog when posting a toot with media that has no alt text. The setting defaults to `true`.
- Implements a modal that will pop up when the user attempts to post the toot, which gives them the opportunity to either go back and add the alt text or to publish anyway.

Screenshot of the new setting:

Screenshot of an checkbox item in the Mastodon user settings that says 'Display reminder to add image descriptions when posting media' with the subtitle 'A reminder will be presented when attempting to post media without an image description'

Screenshot of the current modal that shows up when attempting to publish a toot with no alt text:

A modal dialog in the Mastodon user interface that has a header and subtitle. The header is 'Your attachments are missing a description!' and the subtitle is 'Descriptions are for users with visual or audio impairments that prevent them from seeing or hearing an image, video, or audio file. It's recommended that you provide a description for all attached media files so everyone can enjoy using Mastodon.'. There are two buttons at the bottom of the modal, a purple button that says 'Go back' and a gray button that says 'Publish anyway'.

This modal will show up for all media: Images, video, and audio. It also checks *every* media file in the toot. So if you have 4 images in a toot, but only 3 have alt text, you'll still get this reminder modal.

TODO: Run the internationalization scripts, but I'd to make sure we agree on the copy first.

### CODE REVIEW

File: app/javascript/mastodon/features/ui/components/alt_text_warning_modal.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import Button from '../../../components/button';
+import { closeModal } from '../../../actions/modal';
+
+const makeMapStateToProps = () => {
+  const mapStateToProps = state => ({
+    media: state.getIn(['compose', 'media_attachments']),
+  });
+
+  return mapStateToProps;
+};
```
Comment (connorshea) [inline app/javascript/mastodon/features/ui/components/alt_text_warning_modal.js:14]: This is actually entirely pointless, I don't use the `media` prop at all, but I couldn't figure out how to get rid of it without this component breaking. Anyone know how I can get rid of this?

Comment (graue) [reply app/javascript/mastodon/features/ui/components/alt_text_warning_modal.js:14]: Maybe replace it with `null`? [React-Redux docs say](https://react-redux.js.org/api/connect#connect-parameters):

> If you don't want to subscribe to store updates, pass `null` or `undefined` in place of `mapStateToProps`.

**Comment by Gargron** [comment]
Open to the feature, not eager on the modal window alert. How about adding a red border on the media attachment and making the "No description added" string red instead, as well as disabling the publish button?

**Comment by jorgegamedev** [comment]
As someone who uses description a lot, I personally like it being a modal window, and having a big warning forefront has been has avoided me from making posts without alt texts on another certain social media channel. 

I feel like a red border would still be easy to miss if you're distracted and/or accidentally press the post button. If not a full modal, maybe a warning above the post saying "Your attachments are missing a description!"  similar to the one that appears for direct mentions?

**Comment by connorshea** [comment]
> Open to the feature, not eager on the modal window alert. How about adding a red border on the media attachment and making the "No description added" string red instead, as well as disabling the publish button?

@Gargron I can do that (or the warning like jorge suggested), but there are a few reasons I think a modal works better:

- We're able to communicate to the user the reasons that they should include an image description, because for a lot of users they may not be familiar with the reasoning for/benefits of using alt text. With a modal, we have more space to add to the current message and explain how to write good image descriptions, for example.
- This UX is easy to reproduce on mobile, for the various Mastodon apps.
- If we force them to add an image description, rather than offering an escape hatch of "Publish anyway", realistically many users will just add crappy, useless image descriptions, and that's IMO worse than having none at all.

Also, as suggested the red (regardless of shade) probably wouldn't be high enough contrast to pass web accessibility guidelines, so we'd need to rely primarily on the border styling (or change the way we display the image preview in the compose window entirely):

Screen Shot 2022-11-28 at 5 44 19 PM

**Comment by two-thirty-seven** [comment]
I think the modal is the better option because it trains the user to add alt text - it doesn't force them, but it guides them to fall into the "pit of success".

I would suggest maybe adding some admin-only options to 
a) toggle this feature on and off for their instance and;
b) letting the admins customise the message (solves the internationalization question).

**Comment by orowith2os** [comment]
@SebastianBerlin a concern about forcing it was that the poster might be lazy and add a description that doesn't describe the image at all, which is partially why I said that the feature would be useful alongside https://github.com/mastodon/mastodon/issues/19907

**Comment by graue** [comment]
Also, requiring descriptions itself presents an access issue. Some disabilities can make it not possible to enter descriptions when posting.

**Review by MalloryA — APPROVED** [review]
I tested out this PR on desktop (Firefox/macOS) and mobile (Firefox/iOS).

- I can "publish anyway"
- I can "go back" and add my image description
- Works with one image or two images
- I can turn the setting off and it does not remind me to add a description

I don't have committing privileges, so I can't do anything about it but: lgtm 👍

**Comment by MalloryA** [comment]
> With a modal, we have more space to add to the current message and explain how to write good image descriptions, for example.

I think the copy in the screenshot is great, but I agree a brief explanation of how to write image descriptions would be helpful.

This is a useful page from W3C for image descriptions: https://www.w3.org/WAI/tutorials/images/

I've tried to paraphrase very tersely what they said - as a first suggestion for what instructions should be included in the dialog:

> Images containing text or audio/video content containing spoken words should include a full transcript of the words used and any other essential information. Photographs, illustrations, graphs, and diagrams should include a brief description of what is being presented followed by enough detail that somebody unable to see the image would come away with the same understanding as somebody who can. Information included in the text of a post does not need to be repeated in the description.

File: app/javascript/mastodon/locales/en.json [hunk 2]
```diff
@@ -125,6 +125,10 @@
   "community.column_settings.remote_only": "Remote only",
   "compose.language.change": "Change language",
   "compose.language.search": "Search languages...",
+  "compose_form.alt_text_warning.header": "Your attachments are missing a description!",
+  "compose_form.alt_text_warning.message": "Descriptions are for users with visual or audio impairments that prevent them from seeing or hearing an image, video, or audio file. It's recommended that you provide a description for all attached media files so everyone can enjoy using Mastodon.",
```
Comment (hs4man21) [inline app/javascript/mastodon/locales/en.json:129]: Please update with more inclusive language (e.g. "users who are deaf, hard of hearing, blind or have low vision") per guidance in #20923. Words like "impairment" have negative connotations and are not best practice when talking about folks with disabilities. Thank you!

Comment (connorshea) [reply app/javascript/mastodon/locales/en.json:129]: @hs4man21 I've updated the phrasing, thanks for pointing this out 👍

**Review by garridinsi — APPROVED** [review]
Hi! We implemented this on social.ferrocarril.net and works perfectly! https://social.ferrocarril.net/@garridinsi/109637575107923236

**Comment by mjankowski** [comment]
Closing due to inactivity and merge conflicts ... but doing this (or something like this) do seem useful ... if still interested, can you rebase (some relevant files have changed/moved), fix the lint issues, and reopen/comment?

**Pull request closed without merging**